main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -1138,7 +1139,52 @@ object Typer {
         } yield (matchConstrs.flatten ++ ruleBodyConstrs.flatten, resultTyp, resultPur, resultEff)
 
 
-      case KindedAst.Expression.RestrictableChoose(star, exps0, rules0, tvar, loc) => ??? // TODO RESTR-VARS
+      case KindedAst.Expression.RestrictableChoose(true, exp0, rules0, tvar, loc) => ??? // TODO RESTR-VARS
+
+      case KindedAst.Expression.RestrictableChoose(false, exp0, rules0, tvar, loc) =>
+        // dom(M)
+        val leftSyms = rules0.map(_.pat).map {
+          case KindedAst.RestrictableChoicePattern.Tag(sym, _, _, _) => sym.sym
+        }
+
+        val enumSym = rules0.headOption.getOrElse(throw InternalCompilerException("unexpected empty choose", loc)).pat match {
+          case RestrictableChoicePattern.Tag(sym, pat, tvar, loc) => sym.sym.enumSym
+        }
+
+        val index = root.restrictableEnums(enumSym).index
+        val tparams = root.restrictableEnums(enumSym).tparams
+
+        val targs = (index :: tparams).map {
+          case KindedAst.TypeParam(name, sym, loc) => Type.freshVar(sym.kind, loc.asSynthetic, sym.isRegion)
+        }
+        val enumConstructorKind = Kind.mkArrow(targs.map(_.kind))
+        val enumConstructor = Type.Cst(TypeConstructor.RestrictableEnum(enumSym, enumConstructorKind), loc.asSynthetic)
+
+
+        val enumType = Type.mkApply(enumConstructor, targs, loc.asSynthetic)
+
+        val dom = leftSyms.map {
+          case sym => Type.Cst(TypeConstructor.CaseConstant(sym), loc.asSynthetic): Type
+        }.reduceLeft[Type] {
+          case (acc: Type, tpe: Type) => Type.mkApply(
+            Type.Cst(TypeConstructor.CaseUnion(enumSym), loc.asSynthetic),
+            List(acc, tpe),
+            loc.asSynthetic
+          ): Type
+        }
```

Comment:
I think the above could benefit from more helper functions (especially when moved to a new class).